src/GovernorCountingFractional.sol

```diff
@@ -237,8 +246,101 @@ abstract contract GovernorCountingFractional is Governor {
+     * must be appended to the `params` and included in the signature. I.e., the
+     * params used when constructing the signature would be:
+     *
+     *   abi.encodePacked(againstVotes, forVotes, abstainVotes, nonce)
+     *
+     * See {nonces} and {_castVote} for more information.
+     */
+    function castVoteWithReasonAndParamsBySig(
+        uint256 proposalId,
+        uint8 support,
+        string calldata reason,
+        bytes memory params,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) public virtual override returns (uint256) {
+        // Signature-based fractional voting requires `params` be two full words
+        // in length:
+        //   16 bytes for againstVotes.
+        //   16 bytes for forVotes.
+        //   16 bytes for abstainVotes.
+        //   16 bytes for the signature nonce.
+        // Signature-based nominal voting requires `params` be 0 bytes.
+        require(
+          params.length == 64 || params.length == 0,
+          "GovernorCountingFractional: invalid params for signature-based vote"
+        );
+
+        address voter = ECDSA.recover(
+            _hashTypedDataV4(
+                keccak256(
+                    abi.encode(
+                        EXTENDED_BALLOT_TYPEHASH,
+                        proposalId,
+                        support,
+                        keccak256(bytes(reason)),
+                        keccak256(params)
+                    )
+                )
+            ),
+            v,
+            r,
+            s
+        );
+
+        // If params are zero-length all of the voter's weight will be cast so
+        // we don't have to worry about checking/incrementing a nonce.
+        if (params.length == 64) {
+          // Get the nonce out of the params. It is the last half-word.
+          uint128 nonce;
+          assembly {
+            nonce := and(
+              // Perform bitwise AND operation on the data in the second word of
+              // `params` with a mask of 128 zeros followed by 128 ones, i.e. take
+              // the last 128 bits of `params`.
+              _HALF_WORD_RIGHT,
+              // Load the data from memory at the returned address.
+              mload(
+                // Skip the first 64 bytes (i.e. 0x40):
+                //   32 bytes for the length of the bytes array.
+                //   32 bytes for the first word in the params
+                // Return the memory address for the last word in params.
+                add(params, 0x40)
+              )
+            )
+          }
+
+          require(
+            nonces[voter] == nonce,
+            "GovernorCountingFractional: signature has already been used"
+          );
+
+          nonces[voter]++;
+
+          // Trim params to only keep the first 48 bytes, which are the voting params.
+          assembly {
+            mstore(params, 48)
+          }
+        }
+
+        return _castVote(proposalId, voter, support, reason, params);
```

Comment:
I like named function parameters, even though OZ doesn't use them:

`return _castVote({proposalId: proposalId, account: voter, support: support, reason: reason, params: params});`

Reply:
Probably another one where we'll stick with OZ convention since we're extending it. I also like the named params.

Reply:
Yup, going to stick with OZ's convention, but I also prefer using named parameters